BUG：got error when launch model with vllm-v0.3.2  in multiple GPU mode

### Describe the bug
In order to use Qwen1.5 with vllm  in xinferrence， i upgrade vllm from v0.3.0 to v0.3.2, bug got the following error messsage when launch model with multiple GPU:


**AssertionError: [address=0.0.0.0:10725, pid=58281] daemonic processes are not allowed to have children**
![image](https://github.com/xorbitsai/inference/assets/49679672/0f229d9c-20f8-407f-a674-a875f26c4a12)

but when change the num of GPU to 1 , it works fine


### To Reproduce
To help us to reproduce this bug, please provide information below:

1. Your Python version.  --> python3.11
2. The version of xinference you use. --> v0.8.5
3. Versions of crucial packages. -->  vllm 0.3.2
4. Full stack of the error. --> AssertionError: [address=0.0.0.0:10725, pid=58281] daemonic processes are not allowed to have children
5. Minimized code to reproduce the error.  --> enable vllm and launch xinferrence -> choose 'qwen1.5-chat' -> choose pytorch format and n_gpu > launch model 

![image](https://github.com/xorbitsai/inference/assets/49679672/a89d1e33-e7dc-4960-b55f-a2fe52840555)

Not suer this bug related to the xinferrence of vllm

Confirmed that vllm==0.3.2 has this issue, the full traceback:

``` python
2024-02-22 06:33:22,982 xinference.api.restful_api 155837 ERROR    [address=0.0.0.0:33063, pid=156010] daemonic processes are not allowed to have children
Traceback (most recent call last):
  File "/home/codingl2k1/inference/xinference/api/restful_api.py", line 674, in launch_model
    model_uid = await (await self._get_supervisor_ref()).launch_builtin_model(
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/xoscar/backends/context.py", line 227, in send
    return self._process_result_message(result)
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/xoscar/backends/context.py", line 102, in _process_result_message
    raise message.as_instanceof_cause()
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/xoscar/backends/pool.py", line 657, in send
    result = await self._run_coro(message.message_id, coro)
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/xoscar/backends/pool.py", line 368, in _run_coro
    return await coro
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/xoscar/api.py", line 384, in __on_receive__
    return await super().__on_receive__(message)  # type: ignore
  File "xoscar/core.pyx", line 558, in __on_receive__
    raise ex
  File "xoscar/core.pyx", line 520, in xoscar.core._BaseActor.__on_receive__
    async with self._lock:
  File "xoscar/core.pyx", line 521, in xoscar.core._BaseActor.__on_receive__
    with debug_async_timeout('actor_lock_timeout',
  File "xoscar/core.pyx", line 526, in xoscar.core._BaseActor.__on_receive__
    result = await result
  File "/home/codingl2k1/inference/xinference/core/supervisor.py", line 772, in launch_builtin_model
    await _launch_model()
  File "/home/codingl2k1/inference/xinference/core/supervisor.py", line 736, in _launch_model
    await _launch_one_model(rep_model_uid)
  File "/home/codingl2k1/inference/xinference/core/supervisor.py", line 720, in _launch_one_model
    await worker_ref.launch_builtin_model(
  File "xoscar/core.pyx", line 284, in __pyx_actor_method_wrapper
    async with lock:
  File "xoscar/core.pyx", line 287, in xoscar.core.__pyx_actor_method_wrapper
    result = await result
  File "/home/codingl2k1/inference/xinference/core/utils.py", line 45, in wrapped
    ret = await func(*args, **kwargs)
  File "/home/codingl2k1/inference/xinference/core/worker.py", line 549, in launch_builtin_model
    await model_ref.load()
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/xoscar/backends/context.py", line 227, in send
    return self._process_result_message(result)
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/xoscar/backends/context.py", line 102, in _process_result_message
    raise message.as_instanceof_cause()
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/xoscar/backends/pool.py", line 657, in send
    result = await self._run_coro(message.message_id, coro)
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/xoscar/backends/pool.py", line 368, in _run_coro
    return await coro
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/xoscar/api.py", line 384, in __on_receive__
    return await super().__on_receive__(message)  # type: ignore
  File "xoscar/core.pyx", line 558, in __on_receive__
    raise ex
  File "xoscar/core.pyx", line 520, in xoscar.core._BaseActor.__on_receive__
    async with self._lock:
  File "xoscar/core.pyx", line 521, in xoscar.core._BaseActor.__on_receive__
    with debug_async_timeout('actor_lock_timeout',
  File "xoscar/core.pyx", line 524, in xoscar.core._BaseActor.__on_receive__
    result = func(*args, **kwargs)
  File "/home/codingl2k1/inference/xinference/core/model.py", line 239, in load
    self._model.load()
  File "/home/codingl2k1/inference/xinference/model/llm/vllm/core.py", line 139, in load
    self._engine = AsyncLLMEngine.from_engine_args(engine_args)
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/vllm/engine/async_llm_engine.py", line 625, in from_engine_args
    engine = cls(parallel_config.worker_use_ray,
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/vllm/engine/async_llm_engine.py", line 321, in __init__
    self.engine = self._init_engine(*args, **kwargs)
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/vllm/engine/async_llm_engine.py", line 366, in _init_engine
    return engine_class(*args, **kwargs)
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/vllm/engine/llm_engine.py", line 118, in __init__
    self._init_workers_ray(placement_group)
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/vllm/engine/llm_engine.py", line 286, in _init_workers_ray
    self._run_workers("init_model", cupy_port=get_open_port())
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/vllm/engine/llm_engine.py", line 1014, in _run_workers
    driver_worker_output = getattr(self.driver_worker,
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/vllm/worker/worker.py", line 94, in init_model
    init_distributed_environment(self.parallel_config, self.rank,
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/vllm/worker/worker.py", line 275, in init_distributed_environment
    cupy_utils.init_process_group(
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/vllm/model_executor/parallel_utils/cupy_utils.py", line 90, in init_process_group
    _NCCL_BACKEND = NCCLBackendWithBFloat16(world_size, rank, host, port)
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/cupyx/distributed/_nccl_comm.py", line 70, in __init__
    self._init_with_tcp_store(n_devices, rank, host, port)
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/cupyx/distributed/_nccl_comm.py", line 88, in _init_with_tcp_store
    self._store.run(host, port)
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/site-packages/cupyx/distributed/_store.py", line 100, in run
    p.start()
  File "/home/codingl2k1/.pyenv/versions/3.9.18/lib/python3.9/multiprocessing/process.py", line 118, in start
    assert not _current_process._config.get('daemon'), \
AssertionError: [address=0.0.0.0:33063, pid=156010] daemonic processes are not allowed to have children
```

--- a/setup.cfg
+++ b/setup.cfg
@@ -24,7 +24,7 @@ zip_safe = False
 include_package_data = True
 packages = find:
 install_requires =
-    xoscar>=0.2.1
+    xoscar>=0.3.0
     torch
     gradio>=3.39.0
     pillow
@@ -90,7 +90,7 @@ all =
     einops
     tiktoken
     sentence-transformers>=2.3.1
-    vllm>=0.2.6,<0.3.1 ; sys_platform=='linux'
+    vllm>=0.2.6 ; sys_platform=='linux'
     diffusers
     controlnet_aux
     orjson
@@ -117,7 +117,7 @@ transformers =
     optimum
     peft
 vllm =
-    vllm>=0.2.6,<0.3.1
+    vllm>=0.2.6
 embedding =
     sentence-transformers>=2.3.1
 image =

--- a/xinference/client/tests/test_client.py
+++ b/xinference/client/tests/test_client.py
@@ -610,9 +610,9 @@ def setup_cluster():
             raise RuntimeError("Endpoint is not available after multiple attempts")
 
         yield f"http://localhost:{port}", supervisor_address
-        restful_api_proc.terminate()
+        restful_api_proc.kill()
     finally:
-        local_cluster.terminate()
+        local_cluster.kill()
 
 
 def test_auto_recover(set_auto_recover_limit, setup_cluster):

--- a/xinference/conftest.py
+++ b/xinference/conftest.py
@@ -208,10 +208,11 @@ def setup():
     if not api_health_check(endpoint, max_attempts=10, sleep_interval=5):
         raise RuntimeError("Endpoint is not available after multiple attempts")
 
-    yield f"http://localhost:{port}", supervisor_addr
-
-    local_cluster_proc.kill()
-    restful_api_proc.kill()
+    try:
+        yield f"http://localhost:{port}", supervisor_addr
+    finally:
+        local_cluster_proc.kill()
+        restful_api_proc.kill()
 
 
 @pytest.fixture
@@ -239,10 +240,11 @@ def setup_with_file_logging():
     if not api_health_check(endpoint, max_attempts=3, sleep_interval=5):
         raise RuntimeError("Endpoint is not available after multiple attempts")
 
-    yield f"http://localhost:{port}", supervisor_addr, TEST_LOG_FILE_PATH
-
-    local_cluster_proc.kill()
-    restful_api_proc.kill()
+    try:
+        yield f"http://localhost:{port}", supervisor_addr, TEST_LOG_FILE_PATH
+    finally:
+        local_cluster_proc.kill()
+        restful_api_proc.kill()
 
 
 @pytest.fixture
@@ -290,11 +292,12 @@ def setup_with_auth():
     if not api_health_check(endpoint, max_attempts=10, sleep_interval=5):
         raise RuntimeError("Endpoint is not available after multiple attempts")
 
-    yield f"http://localhost:{port}", supervisor_addr
-
-    local_cluster_proc.kill()
-    restful_api_proc.kill()
     try:
-        os.remove(auth_file)
-    except:
-        pass
+        yield f"http://localhost:{port}", supervisor_addr
+    finally:
+        local_cluster_proc.kill()
+        restful_api_proc.kill()
+        try:
+            os.remove(auth_file)
+        except:
+            pass

--- a/xinference/conftest.py
+++ b/xinference/conftest.py
@@ -210,8 +210,8 @@ def setup():
 
     yield f"http://localhost:{port}", supervisor_addr
 
-    local_cluster_proc.terminate()
-    restful_api_proc.terminate()
+    local_cluster_proc.kill()
+    restful_api_proc.kill()
 
 
 @pytest.fixture
@@ -241,8 +241,8 @@ def setup_with_file_logging():
 
     yield f"http://localhost:{port}", supervisor_addr, TEST_LOG_FILE_PATH
 
-    local_cluster_proc.terminate()
-    restful_api_proc.terminate()
+    local_cluster_proc.kill()
+    restful_api_proc.kill()
 
 
 @pytest.fixture
@@ -292,8 +292,8 @@ def setup_with_auth():
 
     yield f"http://localhost:{port}", supervisor_addr
 
-    local_cluster_proc.terminate()
-    restful_api_proc.terminate()
+    local_cluster_proc.kill()
+    restful_api_proc.kill()
     try:
         os.remove(auth_file)
     except:

--- a/xinference/core/tests/test_metrics.py
+++ b/xinference/core/tests/test_metrics.py
@@ -48,9 +48,9 @@ def setup_cluster():
             raise RuntimeError("Endpoint is not available after multiple attempts")
 
         yield f"http://localhost:{port}", f"http://localhost:{metrics_port}/metrics", supervisor_address
-        restful_api_proc.terminate()
+        restful_api_proc.kill()
     finally:
-        local_cluster.terminate()
+        local_cluster.kill()
 
 
 @pytest.mark.asyncio

--- a/xinference/deploy/docker/cpu.Dockerfile
+++ b/xinference/deploy/docker/cpu.Dockerfile
@@ -21,7 +21,7 @@ ARG PIP_INDEX=https://pypi.org/simple
 RUN python -m pip install --upgrade -i "$PIP_INDEX" pip && \
     pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cpu && \
     pip install -i "$PIP_INDEX" \
-      "xoscar>=0.2.1" \
+      "xoscar>=0.3.0" \
       "gradio>=3.39.0" \
       pillow \
       click \

--- a/xinference/deploy/local.py
+++ b/xinference/deploy/local.py
@@ -132,4 +132,4 @@ def main(
             auth_config_file=auth_config_file,
         )
     finally:
-        local_cluster.terminate()
+        local_cluster.kill()

--- a/xinference/deploy/supervisor.py
+++ b/xinference/deploy/supervisor.py
@@ -98,4 +98,4 @@ def main(
             auth_config_file=auth_config_file,
         )
     finally:
-        local_cluster.terminate()
+        local_cluster.kill()

--- a/xinference/model/llm/pytorch/tests/test_opt.py
+++ b/xinference/model/llm/pytorch/tests/test_opt.py
@@ -116,37 +116,36 @@ async def test_opt_pytorch_model(setup, quantization):
 @pytest.mark.asyncio
 async def test_concurrent_pytorch_model(setup):
     pool = await xoscar.create_actor_pool("127.0.0.1", n_process=1)
-    print(pool.external_address)
-    print(next(iter(pool.sub_processes.keys())))
-    mock_torch_model = MockPytorchModel()
-    model_torch_actor = await xoscar.create_actor(
-        ModelActor,
-        pool.external_address,
-        mock_torch_model,
-        address=next(iter(pool.sub_processes.keys())),
-    )
-    coros = []
-    for _ in range(3):
-        co = model_torch_actor.generate(
-            "Once upon a time, there was a very old computer"
+    async with pool:
+        mock_torch_model = MockPytorchModel()
+        model_torch_actor = await xoscar.create_actor(
+            ModelActor,
+            pool.external_address,
+            mock_torch_model,
+            address=next(iter(pool.sub_processes.keys())),
         )
-        coros.append(co)
-    r = await asyncio.gather(*coros)
-    assert any(r)
+        coros = []
+        for _ in range(3):
+            co = model_torch_actor.generate(
+                "Once upon a time, there was a very old computer"
+            )
+            coros.append(co)
+        r = await asyncio.gather(*coros)
+        assert any(r)
 
-    mock_non_torch_model = MockNonPytorchModel()
-    model_non_torch_actor = await xoscar.create_actor(
-        ModelActor,
-        pool.external_address,
-        mock_non_torch_model,
-        address=next(iter(pool.sub_processes.keys())),
-    )
-    coros = []
-    for _ in range(3):
-        co = model_non_torch_actor.generate(
-            "Once upon a time, there was a very old computer"
+        mock_non_torch_model = MockNonPytorchModel()
+        model_non_torch_actor = await xoscar.create_actor(
+            ModelActor,
+            pool.external_address,
+            mock_non_torch_model,
+            address=next(iter(pool.sub_processes.keys())),
         )
-        coros.append(co)
-    r = await asyncio.gather(*coros)
-    r = [json.loads(i) for i in r]
-    assert not any(r)
+        coros = []
+        for _ in range(3):
+            co = model_non_torch_actor.generate(
+                "Once upon a time, there was a very old computer"
+            )
+            coros.append(co)
+        r = await asyncio.gather(*coros)
+        r = [json.loads(i) for i in r]
+        assert not any(r)

--- a/xinference/model/llm/vllm/core.py
+++ b/xinference/model/llm/vllm/core.py
@@ -13,6 +13,7 @@
 # limitations under the License.
 
 import logging
+import multiprocessing
 import time
 import uuid
 from typing import (
@@ -119,6 +120,7 @@ class VLLMModel(LLM):
 
     def load(self):
         try:
+            import vllm
             from vllm.engine.arg_utils import AsyncEngineArgs
             from vllm.engine.async_llm_engine import AsyncLLMEngine
         except ImportError:
@@ -129,6 +131,14 @@ class VLLMModel(LLM):
             ]
 
             raise ImportError(f"{error_message}\n\n{''.join(installation_guide)}")
+
+        if vllm.__version__ >= "0.3.1":
+            # from vllm v0.3.1, it uses cupy as NCCL backend
+            # in which cupy will fork a process
+            # only for xoscar >= 0.3.0, new process is allowed in subpool
+            # besides, xinference set start method as forkserver for unix
+            # we need to set it to fork to make cupy NCCL work
+            multiprocessing.set_start_method("fork", force=True)
 
         self._model_config = self._sanitize_model_config(self._model_config)
         logger.info(

